File: iq/list/singly/flatten/do.go
// Copyright 2025 Samvel Khalatyan. All rights reserved.

package flatten

import "fmt"

func NewTree(nn ...int) *TreeNode {
	var (
		head *TreeNode
		curr *TreeNode
	)
	for _, n := range nn {
		node := &TreeNode{
			val: n,
		}
		if head == nil {
			head = node
			curr = node
			continue
		}
		curr.next = node
		curr = node
	}
	return head
}

type TreeNode struct {
	val   int
	next  *TreeNode
	child *TreeNode
}

func (tn *TreeNode) Get(val int) *TreeNode {
	for n := tn; n != nil; n = n.next {
		if n.val == val {
			return n
		}
	}
	return nil
}

func (tn *TreeNode) SetChild(child *TreeNode) {
	tn.child = child
}

type Node struct {
	val  int
	next *Node
}

func NewList(nn ...int) *Node {
	var (
		head *Node
		curr *Node
	)
	for _, n := range nn {
		node := &Node{
			val: n,
		}
		if head == nil {
			head = node
			curr = node
			continue
		}
		curr.next = node
		curr = node
	}
	return head
}

func (n *Node) Slice() []int {
	var vv []int
	for n != nil {
		vv = append(vv, n.val)
		n = n.next
	}
	return vv
}

func (n *Node) String() string {
	var vv []int
	for ; n != nil; n = n.next {
		vv = append(vv, n.val)
	}
	return fmt.Sprint(vv)
}

func Do(tn *TreeNode) *Node {
	var (
		head *Node
		tail *Node
	)
	for queue := []*TreeNode{tn}; len(queue) > 0; queue = queue[1:] {
		h, t, q := flattenOneLayer(queue[0])
		switch {
		case head == nil:
			head = h
		default:
			tail.next = h
		}
		tail = t
		queue = append(queue, q...)
	}
	return head
}

func flattenOneLayer(tn *TreeNode) (head *Node, tail *Node, queue []*TreeNode) {
	for ; tn != nil; tn = tn.next {
		n := &Node{val: tn.val}
		switch {
		case head == nil:
			head = n
		default:
			tail.next = n
		}
		tail = n
		if tn.child != nil {
			queue = append(queue, tn.child)
		}
	}
	return head, tail, queue
}

File: iq/17/flatten/do.cc
// Copyright 2026 Samvel Khalatyan. All rights reserved.
//
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

//
// clang-format off-next-line
//go:build ignore

#include "iq/17/flatten/do.h"

#include <functional>
#include <memory>
#include <queue>
#include <utility>

namespace iq::flatten {
namespace {

std::pair<std::shared_ptr<Node>, std::shared_ptr<Node>> toList(
    std::shared_ptr<TreeNode> tree,
    std::function<void(std::shared_ptr<TreeNode>)> visit_child) {
  std::shared_ptr<Node> head;
  std::shared_ptr<Node> tail;
  for (; tree; tree = tree->next) {
    if (tree->child) {
      visit_child(tree->child);
    }
    auto node = std::make_shared<Node>(tree->val);
    if (!head) {
      head = node;
    } else {
      tail->next = node;
    }
    tail = std::move(node);
  }
  return std::make_pair(std::move(head), std::move(tail));
}

}  // namespace

std::shared_ptr<Node> Do([[maybe_unused]] std::shared_ptr<TreeNode> tree) {
  std::shared_ptr<Node> head;
  std::shared_ptr<Node> tail;
  std::queue<std::shared_ptr<TreeNode>> queue;
  auto enque = [&queue](std::shared_ptr<TreeNode> tree_node) {
    queue.emplace(std::move(tree_node));
  };
  for (enque(std::move(tree)); !queue.empty(); queue.pop()) {
    std::shared_ptr<TreeNode> tree_node = queue.front();
    auto [h, t] = toList(tree_node, enque);
    if (!head) {
      head = std::move(h);
    } else {
      tail->next = std::move(h);
    }
    tail = std::move(t);
  }
  return head;
}

}  // namespace iq::flatten
